File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/outtake/Pendulum.java
package org.firstinspires.ftc.teamcode.outtake;

import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.Disabled;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.systems.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.systems.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Outtake.Pendulum.*;

/**
 * Class representing the vertical rotation of the outtake.
 * It extends the ServoPositionMechanism class to control the rotation servo.
 */
public class Pendulum extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servo from.
     */
    public Pendulum(HardwareMap hardwareMap) {
        super(0.0);
        this.hardwareMap = hardwareMap;
    }
    /**
     * Gets the servo associated with the Rotate mechanism.
     * @return an array containing the rotation servo.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Outtake Pendulum 1"), hardwareMap.get(Servo.class, "Outtake Pendulum 2")};
    }
}

/**
 * TeleOp class for testing the Pendulum mechanism manually.
 */
@TeleOp(name = "Outtake Pendulum Test", group = "C Outtake")
@Disabled
class PendulumTest extends ManualMechanismTeleOp {
    public PendulumTest() {
        super(Pendulum::new);
    }
}

/**
 * TeleOp class for testing the Pendulum mechanism positions.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Outtake Pendulum Positions Test", group = "D Outtake")
@Config
@Disabled
class PendulumPositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Pendulum pendulum = new Pendulum(hardwareMap);
        waitForStart();
        while (opModeIsActive()) {
            if (gamepad1.dpad_down) current = transfer;
            else if (gamepad1.dpad_up) current = basket;
            else if (gamepad1.dpad_left) current = bar;
            else if (gamepad1.dpad_right) current = pickup;
            else current += pendulum.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            pendulum.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if(!pendulum.run(packet)) requestOpModeStop();
            telemetry.addData("Position", current);
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/outtake/Pendulum.kt
package org.firstinspires.ftc.outtake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.outtake.Pendulum.Companion.bar
import org.firstinspires.ftc.outtake.Pendulum.Companion.basket
import org.firstinspires.ftc.outtake.Pendulum.Companion.pickup
import org.firstinspires.ftc.outtake.Pendulum.Companion.transfer
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.ServoPositionMechanism

@Config
class Pendulum(hardwareMap: HardwareMap) : ServoPositionMechanism(transfer) {
    companion object {
        /**
         * The position of the outtake pendulum when it is inside of the submersible.
         */
        @JvmField
        @Volatile
        var pickup = 0.834
        /**
         * The position of the outtake pendulum when it is in the transfer position.
         */
        @JvmField
        @Volatile
        var transfer = 0.19
        /**
         * The position of the outtake pendulum when it is in the high chamber position.
         */
        @JvmField
        @Volatile
        var bar = 0.71
        /**
         * The position of the outtake pendulum when it is in the basket position.
         */
        @JvmField
        @Volatile
        var basket = 0.58
    }
    /**
     * Gets the servo associated with the Pendulum mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Outtake Pendulum"])
}
/**
 * TeleOp class for testing the Pendulum mechanism manually.
 */
@TeleOp(name = "Outtake Pendulum Test", group = "C")
@Disabled
private class PendulumTest : ManualMechanismTeleOp(::Pendulum)
/**
 * TeleOp class for testing the positions of the Pendulum mechanism.
 */
@TeleOp(name = "Outtake Pendulum Position Test", group = "D")
@Disabled
private class PendulumPositions : Movement() {
    /**
     * The Pendulum mechanism to be tested.
     */
    override lateinit var system: Pendulum

    override fun systemInit() {
        system = Pendulum(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = pickup })
        actions.add(PressAction(gamepad1::dpad_up) { system.targetPosition = transfer })
        actions.add(PressAction(gamepad1::dpad_left) { system.targetPosition = basket })
        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = bar })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}
